drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,977 @@
+		arg = data->tx_cbs[alloc].arg;
+
+		if (atomic_test_and_clear_bit(data->tx_allocs, alloc)) {
+			if (can_nxp_s32_abort_msg(config,
+					ALLOC_IDX_TO_TXMB_IDX(alloc))) {
+				LOG_ERR("Can't abort message !");
+			};
+
+			function(dev, -ENETDOWN, arg);
+			k_sem_give(&data->tx_allocs_sem);
+		}
+	}
+
+	Canexcel_Ip_EnterFreezeMode(config->instance);
+
+	if (config->phy != NULL) {
+		err = can_transceiver_disable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to disable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
+
+	return 0;
+}
+
+
+static int can_nxp_s32_set_mode(const struct device *dev, can_mode_t mode)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	Canexcel_Ip_ModesType can_nxp_s32_mode = CAN_MODE_NORMAL;
+	bool is_canfd;
+
+	if (data->started) {
+		return -EBUSY;
+	}
+#ifdef CONFIG_CAN_FD_MODE
+	if ((mode & ~(CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY | CAN_MODE_FD)) != 0) {
+#else
+	if ((mode & ~(CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY)) != 0) {
+#endif
+		LOG_ERR("unsupported mode: 0x%08x", mode);
+		return -ENOTSUP;
+	}
+
+	is_canfd = !!(mode & CAN_MODE_FD);
+
+	if (mode & CAN_MODE_LISTENONLY) {
+		can_nxp_s32_mode = CANEXCEL_LISTEN_ONLY_MODE;
+	} else if (mode & CAN_MODE_LOOPBACK) {
+		can_nxp_s32_mode = CANEXCEL_LOOPBACK_MODE;
+	}
+
+	Canexcel_Ip_EnterFreezeMode(config->instance);
+#ifdef CONFIG_CAN_FD_MODE
+	CanXL_SetFDEnabled(config->base_sic, is_canfd, CONFIG_CAN_NXP_S32_BRS);
+#endif
+	CanXL_SetOperationMode(config->base_sic, can_nxp_s32_mode);
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
+
+	return 0;
+}
+
+static int can_nxp_s32_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+
+	__ASSERT(rate != NULL, "Rateptr to get core clock is NULL");
+
+	*rate = config->clock_can;
+
+	return 0;
+}
+
+static int can_nxp_s32_get_max_filters(const struct device *dev, bool ide)
+{
+	ARG_UNUSED(ide);
+
+	return CAN_NXP_S32_MAX_RX;
+}
+
+static int can_nxp_s32_get_max_bitrate(const struct device *dev, uint32_t *max_bitrate)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+
+	*max_bitrate = config->max_bitrate;
+
+	return 0;
+}
+
+static int can_nxp_s32_get_state(const struct device *dev, enum can_state *state,
+						struct can_bus_err_cnt *err_cnt)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	uint32_t sys_status = config->base_sic->SYSS;
+
+	if (state) {
+		if (!data->started) {
+			*state = CAN_STATE_STOPPED;
+		} else {
+			if (sys_status & CANXL_INT_BUSOFF) {
+				*state = CAN_STATE_BUS_OFF;
+			} else if (sys_status & CANXL_INT_PASIVE_ERR) {
+				*state = CAN_STATE_ERROR_PASSIVE;
+			} else if (sys_status & (CANXL_INT_RX_WARNING | CANXL_INT_TX_WARNING)) {
+				*state = CAN_STATE_ERROR_WARNING;
+			} else {
+				*state = CAN_STATE_ERROR_ACTIVE;
+			}
+		}
+	}
+
```

Comment:
You are not filling in the bus error counters in `struct can_bus_err_cnt *err_cnt`?

Reply:
NXP S32 HAL not support error counters, so can't

Reply:
Ok, that's odd. I thought they were required. Please at least set the error counters to zero, then.

Reply:
I updated the counters, set to 0.